mmdet/datasets/crowdhuman.py

```diff
@@ -0,0 +1,151 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import json
+import logging
+import os
+import os.path as osp
+from typing import List, Union
+
+import mmcv
+from mmengine.dataset import BaseDataset
+from mmengine.fileio import FileClient
+from mmengine.logging import print_log
+
+from mmdet.registry import DATASETS
+
+
+@DATASETS.register_module()
+class CrowdHumanDataset(BaseDataset):
+    r"""Dataset for CrowdHuman.
+
+    Args:
+        data_root (str): The root directory for
+        ``data_prefix`` and ``ann_file``.
+        ann_file (str): Annotation file path.
+    """
+
+    METAINFO = {
+        'CLASSES': ('person', ),
+        # PALETTE is a list of color tuples, which is used for visualization.
+        'PALETTE': [(220, 20, 60)]
+    }
+
+    def __init__(self, data_root, ann_file, **kwargs):
+        # id_hw file is an additional annotation information which record the
+        # size of each image. This file is automatically created when you
+        # first load the CrowdHuman dataset by mmdet
+        ann_file_name = os.path.basename(ann_file)
+        if 'train' in ann_file_name:
+            self.id_hw_path = osp.join(data_root, 'id_hw_train.json')
+        elif 'val' in ann_file_name:
+            self.id_hw_path = osp.join(data_root, 'id_hw_val.json')
+        self.id_hw_exist = False
+        if not osp.isfile(self.id_hw_path):
+            print_log(
+                'id_hw file does not exist, prepare to collect '
+                'image height and width...',
+                level=logging.INFO)
+            self.file_client = FileClient()
+            self.id_hw = {}
+        else:
+            self.id_hw_exist = True
+            with open(self.id_hw_path, 'r') as id_hw_file:
+                self.id_hw = json.load(id_hw_file)
+        super().__init__(data_root=data_root, ann_file=ann_file, **kwargs)
+
+    def load_data_list(self) -> List[dict]:
+        """Load annotations from an annotation file named as ``self.ann_file``
+
+        Returns:
+            List[dict]: A list of annotation.
+        """  # noqa: E501
+        with open(self.ann_file, 'r') as file:
+            anno_strs = file.readlines()
+
+        print_log('loading CrowdHuman annotation...', level=logging.INFO)
+        data_list = []
+        prog_bar = mmcv.ProgressBar(len(anno_strs))
+        for i, anno_str in enumerate(anno_strs):
+            anno_dict = json.loads(anno_str)
+            parsed_data_info = self.parse_data_info(anno_dict)
+            data_list.append(parsed_data_info)
+            prog_bar.update()
+        if not self.id_hw_exist:
+            with open(self.id_hw_path, 'w', encoding='utf-8') as file:
```

Comment:
MMDetection's dataset support multiple file client. If the dataset is not stored on disks, such as AWS or Aliyun OSS, this may cause errors.